wagtail_ab_testing/models.py

```diff
@@ -8,20 +12,157 @@ def get_current_for_page(self, page):
     goal_type = models.CharField(max_length=255)
     # TODO Page chooser
     goal_page = models.ForeignKey('wagtailcore.Page', null=True, blank=True, on_delete=models.SET_NULL, related_name='+')
     sample_size = models.PositiveIntegerField()
     status = models.CharField(max_length=20, choices=Status.choices, default=Status.DRAFT)
 
     objects = AbTestManager()
+
+    def finish(self, *, cancel=False):
+        """
+        Finishes the test.
+        """
+        self.status = self.Status.CANCELLED if cancel else self.Status.COMPLETED
+
+        self.save(update_fields=['status'])
+
+    def add_participant(self, variant=None):
+        """
+        Inserts a new participant into the log. Returns the variant that they should be shown.
+        """
+        # Get current numbers of participants for each variant
+        stats = self.hourly_logs.aggregate(
+            control_participants=Sum('participants', filter=Q(variant=self.Variant.CONTROL)),
+            treatment_participants=Sum('participants', filter=Q(variant=self.Variant.TREATMENT)),
+        )
+        control_participants = stats['control_participants'] or 0
+        treatment_participants = stats['treatment_participants'] or 0
+
+        # Create an equal number of participants for each variant
+        if variant is None:
+            if treatment_participants > control_participants:
+                variant = self.Variant.CONTROL
+
+            elif treatment_participants < control_participants:
+                variant = self.Variant.TREATMENT
+
+            else:
+                variant = random.choice([
+                    self.Variant.CONTROL,
+                    self.Variant.TREATMENT,
+                ])
+
+        # Add new participant to statistics model
+        AbTestHourlyLog._increment_stats(self, variant, 1, 0)
+
+        # If we have now reached the required sample size, end the test
+        # Note: we don't care too much that the last few participants won't
+        # get a chance to turn into conversions. It's unlikely to make a
+        # significant difference to the results.
+        # Note: Adding 1 to account for the new participant
+        if control_participants + treatment_participants + 1 >= self.sample_size:
+            self.finish()
+
+        return variant
+
+    def log_conversion(self, variant):
+        """
+        Logs when a participant completed the goal.
+
+        Note: It's up to the caller to make sure that this doesn't get called more than once
+        per participant.
+        """
+        AbTestHourlyLog._increment_stats(self, variant, 0, 1)
+
+
+class AbTestHourlyLog(models.Model):
+    ab_test = models.ForeignKey(AbTest, on_delete=models.CASCADE, related_name='hourly_logs')
+    variant = models.CharField(max_length=9, choices=AbTest.Variant.choices)
+    date = models.DateField()
+    # UTC hour. Values range from 0 to 23
+    hour = models.PositiveSmallIntegerField()
```

Comment:
Points for using the smallest int field available with `PositiveSmallIntegerField`